src/lib.rs

```diff
@@ -0,0 +1,203 @@
+}
+
+fn get_paths_from_dockerfile(content: &str) -> Result<Vec<PathBuf>> {
+    let mut res = vec![];
+
+    for line in get_coalesced_lines_from_dockerfile_content(content.into())? {
+        res.append(&mut get_external_paths_from_dockerfile_line(line)?);
+    }
+
+    Ok(res)
+}
+
+#[cfg(unix)]
+fn metadata_to_u16(metadata: &std::fs::Metadata) -> u16 {
+    let permissions = metadata.permissions();
+    use std::os::unix::fs::PermissionsExt;
+    (permissions.mode() & 0x1ff) as u16
+}
+
+#[cfg(not(unix))]
+fn metadata_to_u16(metadata: &std::fs::Metadata) -> u16 {
+    let permissions = metadata.permissions();
+    // TODO: what else to do on Windows?
+    match (permissions.readonly(), metadata.is_dir()) {
+        (false, false) => 0o444u16,
+        (false, true) => 0o555,
+        (true, false) => 0x666,
+        (true, true) => 0x777,
+    }
+}
+
+pub fn hash(dockerfile_path: &Path, opts: opts::Opts) -> Result<String> {
+    debug!("Opening dockerfile: {}", dockerfile_path.display());
+    let dockerfile_content = std::fs::read_to_string(&dockerfile_path).with_context(|_| {
+        format_err!("Could not read dockerfile: {}", dockerfile_path.display())
+    })?;
+
+    std::env::set_current_dir(&opts.context_path).with_context(|_e| {
+        format_err!(
+            "Couldn't cd to context dir: {}",
+            opts.context_path.display()
+        )
+    })?;
+
+    let rel_ignore_paths: HashSet<_> = opts.ignore_path.clone().into_iter().collect();
+
+    let paths_from_dockerfile = get_paths_from_dockerfile(&dockerfile_content)?;
+
+    for path in &paths_from_dockerfile {
+        info!("Dockerfile depends on: {}", path.display());
+    }
+
+    let mut digests = paths_from_dockerfile
+        .into_iter()
+        .chain(opts.extra_path)
+        .map(|path| {
+            let rdigest = crev_recursive_digest::RecursiveDigest::<blake2::Blake2b, _, _>::new()
+                .additional_data(|entry, writer| {
+                    let metadata = entry.metadata()?;
+                    writer.input(&metadata_to_u16(&metadata).to_be_bytes());
+                    Ok(())
+                })
+                .filter(|entry| {
+                    let rel_path = entry.path().strip_prefix(&path).expect("must be prefix");
+                    !rel_ignore_paths.contains(rel_path)
+                })
+                .build();
+            let digest = rdigest.get_digest_of(&path);
+
+            if let Ok(ref digest) = digest {
+                debug!(
+                    "Partial digest: {} for {}",
+                    base64::encode_config(&digest, base64::URL_SAFE_NO_PAD),
+                    path.display()
+                );
+            }
+
+            digest
+        })
+        .collect::<std::result::Result<Vec<_>, _>>()
+        .map_err(|e| format_err!("{}", e))?;
+
+    digests.append(
+        &mut opts
+            .extra_string
+            .iter()
+            .map(|s| s.as_bytes().to_vec())
+            .collect(),
+    );
```

Comment:
You can avoid the unnecessary collection into a `Vec` by 
```suggestion
    digests.extend(opts.extra_string.iter().map(|s| s.as_bytes().to_vec()));
```